compiler/backend/src/org/jetbrains/kotlin/codegen/ExpressionCodegen.java

```diff
@@ -3476,34 +3480,109 @@ private StackValue generateEquals(
         );
     }
 
-    private StackValue generateEqualityWithInlineClassesBoxing(
+    @NotNull
+    private StackValue genEqualsForInlineClasses(
             @NotNull KtExpression left,
-            @NotNull Type leftType,
             @NotNull KtExpression right,
-            @NotNull Type rightType,
             @NotNull IElementType opToken,
-            @Nullable StackValue pregeneratedSubject
+            @Nullable StackValue pregeneratedSubject,
+            @NotNull KotlinType leftKotlinType,
+            @NotNull KotlinType rightKotlinType,
+            @NotNull Type leftType,
+            @NotNull Type rightType,
+            boolean leftIsUnboxed,
+            boolean rightIsUnboxed
     ) {
-        KotlinType leftKotlinType = kotlinType(left);
-        KotlinType rightKotlinType = kotlinType(right);
-
         StackValue leftValue = genLazyUnlessProvided(pregeneratedSubject, left, leftType, leftKotlinType);
         StackValue rightValue = genLazy(right, rightType, rightKotlinType);
 
         return StackValue.operation(Type.BOOLEAN_TYPE, v -> {
             KotlinType nullableAnyType = state.getModule().getBuiltIns().getNullableAnyType();
+            Label endLabel = new Label();
+            boolean generateEndLabel = false;
+            boolean flipComparison = opToken == KtTokens.EXCLEQ || opToken == KtTokens.EXCLEQEQEQ;
+
+            // Don't call equals-impl0 if the class file version is too low to support it.
+            boolean useUnboxedEquals = leftIsUnboxed && JvmCodegenUtil.typeHasSpecializedInlineClassEquality(leftKotlinType);
 
             leftValue.put(leftType, leftKotlinType, v);
-            StackValue.coerce(leftType, leftKotlinType, AsmTypes.OBJECT_TYPE, nullableAnyType, v);
+            //noinspection SuspiciousNameCombination
+            Type afterTopType = leftType;
+            if (!useUnboxedEquals) {
+                StackValue.coerce(leftType, leftKotlinType, OBJECT_TYPE, nullableAnyType, v);
+                afterTopType = OBJECT_TYPE;
+            }
 
             rightValue.put(rightType, rightKotlinType, v);
-            StackValue.coerce(rightType, rightKotlinType, AsmTypes.OBJECT_TYPE, nullableAnyType, v);
+            //noinspection SuspiciousNameCombination
+            Type topType = rightType;
+            if (!useUnboxedEquals || !rightIsUnboxed) {
```

Comment:
Interesting that we always box the RHS if the LHS is boxed. Maybe we could swap the arguments with `AsmUtil.swap` and call `equals-impl`, to avoid boxing in cases like this?
```
inline class Z(val s: Long)

fun test(z1: Z?, z2: Z): Boolean {
    //  current:   Intrinsics.areEqual(z1, Z.box-impl(z2))
    // proposed:   Z.equals-impl(z2, z1)
    return z1 == z2
}
```

Reply:
I implemented this in the beginning! Unfortunately, it's not a sound optimization, see [equalsCallsLeftArgument.kt](https://github.com/JetBrains/kotlin/pull/2539/files#diff-85b28675505ffc8ed17d13bf53dba531).

Reply:
That is an awesome test! You're right, it's incorrect in the general case.

I wonder if it makes sense to perform this optimization if it's statically known that the RHS is an _inline_ class, where we know that the equality is symmetric? That would remove boxing from the most popular case, where we compare values of the same inline class type, without changing behavior of others.

This approach could supposedly backfire only in a situation where the RHS's inline class is independently recompiled to a non-inline class with non-symmetric `equals`, and behavior of the `==` check will be different at runtime compared to if it was recompiled. I'm not sure if we need to tolerate this though, because seeing the word "inline", a user _could_ be required to accept that any ABI-affecting change will need to cause clients to recompile. I think this is another design question which I'll raise internally...